python/rainbow/simulators/prox_soft_bodies/gauss_seidel_parallel.py

```diff
@@ -0,0 +1,381 @@
+import numpy as np
+import numba as nb
+import networkx as nx
+from collections import defaultdict
+from rainbow.util.timer import Timer
+
+
+def _build_contact_graph(J):
+    """ Create the contact graph based on the contact jacobi matrix.
+        The vertices of the graph are the contact points, and the edges of the graph are the value of the contact jacobi matrix is not zero between two contact points.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        graph: The contact graph.
+    """
+    G = nx.Graph()
+    K = J.shape[0] // 4
+    G.add_nodes_from(np.arange(K))
+    sources, targets = _get_edges(J)
+    for s, t in zip(sources, targets):
+        G.add_edge(s, t) 
+    
+    return G
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def _get_edges(J):
+    """ Get the edges of the contact graph.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        (ArrayLike, ArrayLike): The source and target of the edges.
+    """
+    K = J.shape[0] // 4
+    step_size = 24
```

Comment:
Add comment that explains where 4 and 24 come from

Reply:
I am not so sure what step size is here. Is it the number of non-zeros in a row of J? K appears to be the number of row blocks.

I guess any E-E or V-F contact would involve at most 4 nodes, a velocity vector is a 3D vector, so I would expect 12 non-zeros per row, not 24.